Date: Apr 27 16:25:26 2018

From: 38bf35e06693180bd9002d048f0bb7f9f933e02e Oct 11 16:40:18 2018

Subject: [PATCH] RDK-22717 : CPE side code development for stunnel based reverse ssh

        added socat and stunnel packages restricted socat connection capability to loopback ips
	127.0.0.1 and [::1]
        Updated for socat 1.8.1.1 (_xiogetaddrinfo new signature).

Source: COMCAST


Upstream-Status: Pending

Signed-off-by:  Pritam Ghanghas <pritam_ghanghas@comcast.com>
---

diff --git a/xio-ip.c b/xio-ip.c
--- a/xio-ip.c
+++ b/xio-ip.c
@@ -455,6 +455,20 @@ int _xiogetaddrinfo(const char *node, const char *service,
 		   int family, int socktype, int protocol,
 		   struct addrinfo **res, const int ai_flags[2]) {
    char *numnode = NULL;
+   // this is safest place to limit the ip arguments to take only loopback address
+   // any other place we run the risk of missing some argument cobination,
+   // any ip or name will come here for resolution. We will drop names that require
+   // resoution and IPs other than loopback. Only digits for ipv4 and ipv6 loopback
+   if (!node) {
+       printf("\r\nHardened socat: we don't allow any IP, please use 127.0.0.1 or [::1], we don't even allow localhost\r\n\n");
+       exit(1);
+   }
+   if (strcmp(node, "127.0.0.1") && strcmp(node, "[::1]")) {
+       printf("\r\nHardened socat: Supplied hostname: %s not supported, we don't allow external IPs.\n" \
+                  "                Please use 127.0.0.1 or [::1], we don't even allow localhost\r\n\n", node);
+       exit(1);
+   }
+
    size_t nodelen;
 #if HAVE_GETADDRINFO
    struct addrinfo hints = {0};
